cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java

```diff
@@ -115,49 +117,61 @@ protected void run() throws Exception {
 
     ApplicationSpecification appSpec = program.getApplicationSpecification();
 
-    Iterator<ScheduleProgramInfo> iterator = workflowSpec.getActions().iterator();
+    Iterator<WorkflowNode> iterator = workflowSpec.getNodes().iterator();
     int step = 0;
     while (running && iterator.hasNext()) {
       WorkflowActionSpecification actionSpec;
-      ScheduleProgramInfo actionInfo = iterator.next();
-      switch (actionInfo.getProgramType()) {
-        case MAPREDUCE:
-          MapReduceSpecification mapReduceSpec = appSpec.getMapReduce().get(actionInfo.getProgramName());
-          actionSpec = new DefaultWorkflowActionSpecification(new ProgramWorkflowAction(mapReduceSpec.getName(),
-                                                                                        mapReduceSpec.getName(),
-                                                 SchedulableProgramType.MAPREDUCE));
+      WorkflowNode node = iterator.next();
+      switch (node.getType()) {
+        case ACTION:
+          WorkflowActionNode actionNode = (WorkflowActionNode) node;
```

Comment:
I don't like big switch case like this. Better encapsulate this action to be taken into a method. e.g.

```java
switch (node.getType()) {
  case ACTION:
    executeAction((WorkflowActionNode) node);
  case FORK:
    executeFork((WorkflowForkNode) node);
  ....
}
```